deeprankcore/trainer.py

```diff
@@ -443,18 +441,65 @@ def configure_optimizers(self, optimizer = None, lr: float = 0.001, weight_decay
+                your task (classification or regression). All loss functions
+                from torch.nn.modules.loss are listed as belonging to either
+                category (or to neither) and an exception is raised if an invalid
+                loss function is chosen for the set task.
+                Default for regression: MSELoss
+                Default for classification: CrossEntropyLoss
+            override_invalid (bool): If True, loss functions that are considered
+                invalid for the task do no longer automaticallt raise an exception.
+                Defaults to False.
         """
 
+        default_regression_loss = nn.MSELoss
+        default_classification_loss = nn.CrossEntropyLoss
+
+        default_loss_info = (f'No loss function provided, the default loss function for {self.task} tasks is used: {lossfunction}')
+        custom_loss_warning = ( f'The provided loss function ({lossfunction}) is not part of the default list.\n\t' +
+                                f'Please ensure that this loss function is appropriate for {self.task} tasks.\n\t')
+        invalid_loss_error = (f'The provided loss function ({lossfunction}) is not appropriate for {self.task} tasks.\n\t' + 
+                               'If you want to use this loss function anyway, set override_invalid to True.')
+        override_warning = (f'The provided loss function ({lossfunction}) is not appropriate for {self.task} tasks.\n\t' + 
+                            'You have set override_invalid to True, so the training will run with this loss function nonetheless.\n\t' +
+                            'This will likely cause other errors or exceptions down the line.')
+
+        def _invalid_loss():
+            if override_invalid:
+                _log.warning(override_warning)
+            else:
+                _log.error(invalid_loss_error)
+                raise ValueError(invalid_loss_error)
+
+        # check for custom/invalid loss functions
+        if lossfunction in losses.other_losses:
+            _invalid_loss()
+        elif lossfunction not in (losses.regression_losses + losses.classification_losses):
+            custom_loss = True
+        else:
+            custom_loss = False
+
+        # set regression loss
         if self.task == targets.REGRESS:
-            self.loss = MSELoss()
+            if lossfunction is None:
+                lossfunction = default_regression_loss
+                _log.info(default_loss_info)
+            else:
+                if custom_loss:
+                    _log.warning(custom_loss_warning)
+                elif lossfunction not in losses.regression_losses:
+                    _invalid_loss()
+            self.lossfunction = lossfunction()
 
         elif self.task == targets.CLASSIF:
             # Assign weights to each class
-            self.weights = None
             if self.class_weights:
                 targets_all = []
                 for batch in self.train_loader:
```

Comment:
@gcroci2 , I found an error that is due to PR #311 but wasn't detected there. It only shows up if `class_weights` is set to True, which apparently wasn't the case in any previous test.

Basically, if `class_weights` is True, then it needs to have a `train.loader` defined. However, now the `train_loader` is only defined during training, which happens after the loss function is set.

Any suggestion how to resolve this without reverting to defining `train_loader` in init after all?

Reply:
I've run the tests locally but I got many errors, even in tests that do not user class_weights parameter. Did you notice that as well? Solved those tests, I have an idea about how to move stuff without reinserting the loaders in the init.

Reply:
> I've run the tests locally but I got many errors, even in tests that do not user class_weights parameter. Did you notice that as well? Solved those tests, I have an idea about how to move stuff without reinserting the loaders in the init.

Yeah, I think I added another bug last evening when I was working late/tired. Those should be fixed now and locally I only have the one with weights failing. Also the build is now only failing on the one test.